domain/network/state/linklayer_merge.go

```diff
@@ -0,0 +1,719 @@
+)
+
+// mergeLinkLayerDevice is a subset of linklayerdevice.LinkLayerDevice
+// that is used to merge the existing link layer devices with the
+// incoming ones.
+//
+// It contains only the fields that are used to identify and merge the devices
+type mergeLinkLayerDevice struct {
+	UUID       string
+	Name       string
+	MACAddress string
+	ProviderID string
+	Type       corenetwork.LinkLayerDeviceType
+	Addresses  []mergeAddress
+}
+
+// mergeAddress is a subset of ipaddress.IPAddress that is used to merge
+// the existing addresses with the incoming ones.
+//
+// It contains only the fields that are used to identify and merge the addresses
+type mergeAddress struct {
+	UUID       string
+	Value      string
+	ProviderID string
+}
+
+// mergeLinkLayerDevicesChanges contains the changes to be applied to the
+// link layer devices.
+type mergeLinkLayerDevicesChanges struct {
+	// toAdd maps provider IDs to LinkLayerDeviceUUIDs to be added
+	// in provider_link_layer_device.
+	toAdd map[string]string
+	// ToRemove are the provider IDs to remove from provider_link_layer_device
+	toRemove []string
+	// toRelinquish is a list of LinkLayerDeviceUUIDs to
+	// relinquish to the machine, i.e., set all their addresses origin to machine.
+	toRelinquish []string
+	// newDevices are the incoming devices that did not match any we already
+	// have in state.
+	newDevices []mergeLinkLayerDevice
+}
+
+// mergeAddressesChanges contains the changes to be applied to the
+// addresses.
+type mergeAddressesChanges struct {
+	// toAdd maps provider IDs to ip_address UUID to be added
+	// in provider_link_layer_device.
+	toAdd map[string]string
+	// ToRemove are the provider IDs to remove from provider_ip_address
+	toRemove []string
+	// toRelinquish are a list of ip_address to
+	// relinquish to machine, i.e., set their origin to machine.
+	toRelinquish []string
+}
+
+// MergeLinkLayerDevice merges the existing link layer devices with the
+// incoming ones.
+func (st *State) MergeLinkLayerDevice(
+	ctx context.Context,
+	machineUUID string,
+	incoming []network.NetInterface,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	return db.Txn(
+		ctx, func(ctx context.Context, tx *sqlair.TX) error {
+			nodeUUID, err := st.getMachineNetNodeUUID(ctx, tx,
+				machineUUID)
+			if err != nil {
+				return errors.Errorf(
+					"getting net node UUID for machine %q: %w",
+					machineUUID, err,
+				)
+			}
+
+			existingDevices, err := st.getExistingLinkLayerDevices(
+				ctx, tx, nodeUUID,
+			)
+			if err != nil {
+				return errors.Errorf(
+					"getting existing link layer devices for node %q: %w",
+					nodeUUID, err,
+				)
+			}
+
+			if len(existingDevices) == 0 {
+				// Noop
+				st.logger.Debugf(ctx, "no existing devices, "+
+					"ignoring %d incoming device for net node %q",
+					len(incoming), nodeUUID)
+				return nil
+			}
+
+			normalized, namelessHWAddrs,
+				err := st.normalizeLinkLayeredDevices(ctx,
```

Comment:
This file reads better if each of these methods is located in order beneath this one.